### PR TITLE
Support both use_calc_stream and sync_op in collective communication API

### PR DESCRIPTION
**Pull request by HermitSun**
### PR types

New features
### PR changes

APIs
### Describe

In the new communication library, we designed ProcessGroup to manage different communication group. Inside each process_group has its own stream which all communications in this group will be done on this stream. For high level API, like `distributed.all_reduce`, we use `use_calc_stream` to indicate whether this operation is sync or not. Notice that frequently add unnecessary cuda events may lead to low performance on some model. In order to achieve high performance, this pr add a new API name distributed.stream.all_reduce. This new API provided `use_calc_stream` and `sync_op` both.

- `sync_op`, indicate whether communication is sync or not.
- `use_calc_stream`, do communicate on calc stream, save the time of switching stream. Only work when `sync_op` is true.

对应文档的中文api修改，见 https://github.com/PaddlePaddle/docs/pull/5237

TODO：
- [x] reduce
- [x] broadcast
- [x] all_reduce
- [x] all_gather
- [x] alltoall
- [x] alltoall_single
- [x] sendrecv
- [x] scatter
- [x] reduce_scatter
- [x] _reduce_scatter_base

### CODE REVIEW

File: python/paddle/distributed/communication/__init__.py [hunk 1]
```diff
@@ -12,4 +12,4 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-__all__ = ["stream"]
+__all__ = ["comm_utils", "stream"]
```
Comment (LiYuRio) [inline python/paddle/distributed/communication/__init__.py]: 这个先不用加，要被干掉了

File: paddle/fluid/pybind/distributed_py.cc [hunk 2]
```diff
@@ -621,6 +878,223 @@ void BindDistributed(py::module *m) {
...
+              py::arg("out"),
+              py::call_guard<py::gil_scoped_release>())
+
+          .def(
+              "alltoall_single_on_calc_stream",
+              [](distributed::ProcessGroupStream &self,
+                 py::handle py_in_tensor,
+                 py::handle py_out_tensor,
+                 std::vector<int64_t> in_sizes,
+                 std::vector<int64_t> out_sizes) {
```
Comment (LiYuRio) [inline paddle/fluid/pybind/distributed_py.cc]: 这里的vector用引用可以避免复制

File: paddle/fluid/pybind/distributed_py.cc [hunk 3]
```diff
@@ -480,6 +553,34 @@ void BindDistributed(py::module *m) {
               py::arg("out_sizes"),
               py::call_guard<py::gil_scoped_release>())
 
+          .def(
+              "alltoall_single",
+              [](distributed::ProcessGroup &self,
+                 py::handle py_in_tensor,
+                 py::handle py_out_tensor,
+                 std::vector<int64_t> in_sizes,
+                 std::vector<int64_t> out_sizes,
```
Comment (LiYuRio) [inline paddle/fluid/pybind/distributed_py.cc]: 同下

File: python/paddle/distributed/communication/__init__.py (hunk 1 above)
Comment (HermitSun) [reply python/paddle/distributed/communication/__init__.py]: done

File: paddle/fluid/pybind/distributed_py.cc (hunk 2 above)
Comment (HermitSun) [reply paddle/fluid/pybind/distributed_py.cc]: done

File: paddle/fluid/pybind/distributed_py.cc (hunk 3 above)
Comment (HermitSun) [reply paddle/fluid/pybind/distributed_py.cc]: done

**Review by LiYuRio — DISMISSED** [review]

File: paddle/fluid/distributed/collective/Utils.h [hunk 4]
```diff
@@ -33,17 +45,39 @@ struct SplitDenseTensor {
...
 
 #ifdef PADDLE_WITH_CUSTOM_DEVICE
+template <typename T>
+struct ConcatDenseTensor<platform::CustomDeviceContext, T> {
+  void operator()(const platform::CustomDeviceContext *context,
+                  const std::vector<phi::DenseTensor> &in,
+                  phi::DenseTensor *out,
+                  int axis = 0) {
+    const auto &place = context->GetPlace();
+    phi::DeviceGuard guard(place);
```
Comment (LiYuRio) [inline paddle/fluid/distributed/collective/Utils.h]: 这个guard应该不需要，guard遵循RAII，保证在这个作用域内kernel执行的gpu设备和context一致，但目前一个进程对应一个设备这里应该不会变？

File: paddle/fluid/pybind/distributed_py.cc [hunk 5]
```diff
@@ -600,6 +833,37 @@ void BindDistributed(py::module *m) {
...
+                 int rank_id) {
+                auto in_tensor = CastPyArg2Tensor(py_in_tensor.ptr(), 0);
+                auto out_tensor = CastPyArg2Tensor(py_out_tensor.ptr(), 0);
+                auto in_dense = std::dynamic_pointer_cast<phi::DenseTensor>(
+                    in_tensor.impl());
+                auto out_dense = std::dynamic_pointer_cast<phi::DenseTensor>(
+                    out_tensor.impl());
+                std::vector<phi::DenseTensor> in_tensors = {*in_dense};
+                std::vector<phi::DenseTensor> out_tensors = {*out_dense};
+                int numel = (*in_dense).numel();
```
Comment (LiYuRio) [inline paddle/fluid/pybind/distributed_py.cc]: 这里用int，在数据量比较大的时候可能会发生截断，numel本身返回的是int64类型

Comment (HermitSun) [reply paddle/fluid/pybind/distributed_py.cc]: 这是他们写的，我解决冲突的时候估计这块被标记成修改了……要不我把这些`int numel`都改成int64吧

File: paddle/fluid/distributed/collective/Utils.h (hunk 4 above)
Comment (HermitSun) [reply paddle/fluid/distributed/collective/Utils.h]: 那，要不然先删了？

Comment (LiYuRio) [reply paddle/fluid/distributed/collective/Utils.h]: 好

File: paddle/fluid/pybind/distributed_py.cc (hunk 5 above)
Comment (LiYuRio) [reply paddle/fluid/pybind/distributed_py.cc]: 改一改

Comment (HermitSun) [reply paddle/fluid/pybind/distributed_py.cc]: done

File: paddle/fluid/distributed/collective/Utils.h (hunk 4 above)
Comment (HermitSun) [reply paddle/fluid/distributed/collective/Utils.h]: done

**Review by LiYuRio — APPROVED** [review]

**Review by gongweibao — APPROVED** [review]
LGTM

**Review by liuTINA0907 — APPROVED** [review]

**Review by XieYunshen — APPROVED** [review]
LGTM 
单测超时时间设置

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**